### PR TITLE
MM-26927: Use a recovery middleware

### PR DESCRIPTION
**Pull request by agnivade**
#### Ticket link

https://mattermost.atlassian.net/browse/MM-26927

### CODE REVIEW

**Comment by agnivade** [comment]
Why yes of course :) Thanks for the push.

File: server/server.go [hunk 1]
```diff
@@ -152,6 +153,17 @@ func (s *Server) RefreshMembers() {
 	s.OrgMembers = members
 }
 
+func (s *Server) withRecovery(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			if x := recover(); x != nil {
+				mlog.Error("recovered from a panic", mlog.String("url", r.URL.String()), mlog.Any("error", x))
```
Comment (isacikgoz) [inline server/server.go:160]: Shall we also print the request body and header? It would be great to see which request caused the panic :)

Comment (agnivade) [reply server/server.go:160]: I think that would pollute the logs too much. And plus you need another NopCloser to read the body again.

Header is unlikely to contain any helpful information. The thing to check would be the body. I am guessing that the panic trace should be enough for us to figure out the bug.

Comment (isacikgoz) [reply server/server.go:160]: I see, let's continue with this then. 👍

**Comment by agnivade** [comment]
@metanerd - Please deploy at your leisure.

**Comment by metanerd** [comment]
@agnivade done

**Pull request merged**